integtest/src/main/java/org/corfudb/universe/cluster/docker/DockerCluster.java

```diff
@@ -0,0 +1,200 @@
+
+    @Getter
+    private final ClusterParams clusterParams;
+
+    private final DockerClient docker;
+    private final DockerNetwork network = new DockerNetwork();
+
+    private final ImmutableMap<String, Service> services;
+    private final Optional<String> clusterId;
+    private final Thread shutdownHook;
+
+    @Builder
+    public DockerCluster(ClusterParams clusterParams, DockerClient docker, ImmutableMap<String, Service> services,
+                         Optional<String> clusterId) {
+        this.clusterParams = clusterParams;
+        this.docker = docker;
+        this.services = services;
+        this.clusterId = clusterId;
+
+        shutdownHook = new Thread(this::shutdown);
+        Runtime.getRuntime().addShutdownHook(shutdownHook);
+    }
+
+
+    /**
+     * Deploy a cluster according to provided cluster parameter, docker client, docker network, and other cluster
+     * components.
+     * The instances of this class are immutable. In other word, when the state of an instance is changed a new
+     * immutable instance is provided.
+     *
+     * @return New immutable instance of docker cluster would be returned.
+     * @throws ClusterException this exception will be thrown if deploying a cluster is not successful
+     */
+    @Override
+    public DockerCluster deploy() throws ClusterException {
+        log.info("Deploying cluster");
+
+        if (!clusterId.isPresent()) {
+            network.setup();
+        }
+
+        Map<String, Service> servicesSnapshot = createAndDeployServices();
+
+        Optional<String> newClusterId = clusterId;
+        if (!newClusterId.isPresent()) {
+            newClusterId = Optional.of(UUID.randomUUID().toString());
+        }
+
+        DockerCluster cluster = DockerCluster.builder()
+                .clusterParams(clusterParams)
+                .docker(docker)
+                .services(ImmutableMap.copyOf(servicesSnapshot))
+                .clusterId(newClusterId)
+                .build();
+
+        Runtime.getRuntime().removeShutdownHook(shutdownHook);
+
+        return cluster;
+    }
+
+    @Override
+    public void shutdown() throws ClusterException {
+        log.info("Shutdown docker cluster: {}", clusterId);
+
+        // gracefully stop all services
+        services.values().forEach(service -> {
+            try {
+                service.stop(clusterParams.getTimeout());
+            } catch (Exception ex) {
+                log.info("Can't stop service: {}", service.getParams().getName());
+            }
+        });
+
+        // Kill all docker containers
+        clusterParams.getServices().keySet().forEach(serviceName -> {
+            ServiceParams<ServerParams> serviceParams = clusterParams.getService(serviceName);
+
+            serviceParams.getNodes().forEach(serverParams -> {
+                try {
+                    docker.killContainer(serverParams.getGenericName());
+                } catch (DockerException | InterruptedException e) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Either re-interrupt this method or rethrow the "InterruptedException". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2142)